allennlp/experiments/drivers/train_driver.py

```diff
@@ -0,0 +1,182 @@
+
+    @overrides
+    def run(self) -> None:
+        epoch_counter = 0
+        # Resume from serialization path if it contains a saved model.
+        if self._serialization_prefix is not None:
+            if any(["model_state_epoch_" in x
+                    for x in os.listdir(self._serialization_prefix)]):
+                epoch_counter = self._restore_checkpoint()
+
+        for epoch in range(epoch_counter, self._num_epochs):
+            train_loss = 0.0
+            val_loss = 0.0
+
+            # Set the model to "train" mode.
+            self._model.train()
+            train_generator = self._iterator(self._train_dataset, num_epochs=1)
+            for batch in tqdm.tqdm(train_generator):
+                tensor_batch = arrays_to_variables(batch, self._cuda_device)
+                self._optimizer.zero_grad()
+                output_dict = self._model.forward(**tensor_batch)
+                try:
+                    loss = output_dict["loss"]
+                    loss.backward()
+                    train_loss += loss.data.numpy()
+                except KeyError:
+                    raise ConfigurationError("The model you are trying to optimize does not contain a"
+                                             " 'loss' key in the output of model.forward(inputs).")
+                # TODO(Mark): Do we care about optimizers which require a closure?
+                self._optimizer.step()
+
+            # Switch to evaluation mode.
+            if self._validation_dataset is not None:
+                self._model.eval()
+                val_generator = self._iterator(self._validation_dataset, num_epochs=1)
+                for batch in tqdm.tqdm(val_generator):
+                    tensor_batch = arrays_to_variables(batch, self._cuda_device)
+                    val_output_dict = self._model.forward(tensor_batch)
+                    loss = val_output_dict["loss"]
+                    val_loss += loss.data.numpy()
+
+            # TODO(Mark): Add user specified metrics here, maybe a "metrics" key?
+            logger.info("Training Loss: %3f    Validation Loss: %3f ", train_loss, val_loss)
+            if self._serialization_prefix:
+                self._save_checkpoint(epoch)
+
+    def _save_checkpoint(self,
+                         epoch: int,
+                         is_best: Optional[bool] = None) -> None:
+        model_path = os.path.join(self._serialization_prefix, "model_state_epoch_{}.th".format(epoch))
+        model_state = self._model.state_dict()
+        torch.save(model_state, model_path)
+
+        training_state = {'epoch': epoch, 'optimizer': self._optimizer.state_dict()}
+        torch.save(training_state, os.path.join(self._serialization_prefix,
+                                                "training_state_epoch_{}.th".format(epoch)))
+        if is_best:
+            shutil.copy(model_path, os.path.join(model_path, "best.th"))
+
+    def _restore_checkpoint(self) -> int:
```

Comment:
A short docstring would be helpful here, even though it's a private method.  There are a few choices you make in doing this, and you have a return value that's not obvious from the method signature.